File: J750_to_Ultra_compiler.pl
use feature qw(say);#加载perl的新特性

$outputfolder = "UltraFlex";
mkdir $outputfolder, 0755 or warn "Folder create failed." if ( !-e $outputfolder ); 
$filepath = "J750_Pattern\\";#"input\\"; Jesse@20180411: change J750 .atp folder
opendir FOLDER, $filepath or die "cannot find $filepath";
$is_scan_pat=0; #check the pattern include scan or not

#use for save opcode and opcode located in pattern
$file_opcode_list="opcode_list.txt";
$file_opcode_in_pattern="opcode_in_pattern.txt";
unlink $file_opcode_list       if (-e $file_opcode_list);
unlink $file_opcode_in_pattern if (-e $file_opcode_in_pattern);
my @opcodelist;

#iterate all the files in input folder, and process all the .atp files
my @filelist = readdir(FOLDER);
foreach my $file (@filelist) 
{
	if ( $file =~ /\.atp$/ ) 
	{
		my $filename = $filepath . $file;
		&findopcode( $filename, "$outputfolder\\",$file, &getVectorNum($filename) );   #convert J750 .atp to UltraFlex .atp
		&UltraFlexPatternCompiler ("$outputfolder\\".$file); # compile Ultrafelx ASCII pattern(.atp) to binary pattern(.pat)
	}
}#foreach $file (@filelist)

#couting vector line numbers sub getVectorNum { open READVECTOR,  "<",  $_[0]; 
my $vector_counts=0; my $originStr; while ( defined( $originStr = <READVECTOR> ) 
) { print ''; $vector_counts++ if ($originStr=~/>\s/); } close READVECTOR; 
return $vector_counts; }#getVectorNum

#main function, convert J750 .atp to UltraFlex .atp
sub findopcode
{
	open READ,  "<",  $_[0]; #open J750 .atp for read
	open WRITE, ">", $_[1] . $_[2]; #open a new .atp to convert/write J750 .atp to UlytraFlex
	open WRITE_opcode, ">> $file_opcode_list";
	open WRITE_opcode_location, ">> $file_opcode_in_pattern";
	my $tempt_opcode="";
	
	my $patternnumber=0;
	my $str_line;
	my $outofcolumn = 1;
	my $result="";
	my $originStr="";
	my $orginal_j750;
	my $before="";
	my $pinlist="";
	my $linenumber=0;
	my $subfilename=$_[2];
	my $vector_num=$_[3]; #get the vector counts to determine VM or SRM module.
	$is_scan_pat=0; #check the pattern include scan or not
	$subfilename=~s/\.atp$//;
	$subfilename=~s/-/_/;#replace all illegal characters to support vector module name
	say "$_[0]   Please wait";
	print WRITE_opcode_location "--------------$_[0]--------------\n";
#	my $label_line="";
	my @pattern_header_commets;
	my $flag_header_commets=-1;#indicate the pattern_header_commets
	while ( defined( $originStr = <READ> ) ) 
	{
		chomp($originStr);
		$result="";
		$linenumber++;
		$originStr =~ s/^\s+|\s+$//;
		$orginal_j750=$originStr;
		$str_line=$originStr;
		$tempt_opcode="";#clean tempt opcode
		next if ($originStr eq "");#skip empty line
		
		#process the pattern_header_commets, move the header comments in to vector module
		if($flag_header_commets==-1)
		{ 
			while ( $originStr =~ /^\/\//)
			{
				push(@pattern_header_commets, $originStr);
				$originStr = <READ>;
				chomp($originStr);
				$linenumber++;
				next;
			}
			$flag_header_commets=1;
		}
		
		$is_scan_pat=1 if ($originStr=~/scan_pins\s=/i); #check pattern contain scan or not; scan_pins = {

		#1)check the line is comment or not, if comments just go to write original context directly
		goto WRITE_DIRECTLY if ( $str_line =~ /^\/\/|\/\*/);
		#not comments, do the below process

		#2)check the pattern has scan or not
		if($str_line=~/(\w*):\d,/i) {print WRITE $1 .",\n"; next;} #scan pin setup

		#3)process the vector module header
		if($str_line=~/^vector\s*\(\s*\$tset\s*,/)
		{
			$str_line=~/^vector\s*\(\s*\$tset\s*,(?<pinlist>.*)\)/;
			$pinlist=$+{pinlist};
			#check for VM or SRM(vector line <64>)
			if($vector_num>64)	{print WRITE "vm_vector $subfilename \n";}
			else				{print WRITE "srm_vector $subfilename \n";}
			print WRITE "(\$tset,$pinlist)\n";
			##print WRITE "vm_vector $subfilename (\$tset,$pinlist)\n";
			next;
		}
		#vector context satrt, the pattern_header_commets will be write
		if($str_line=~/^{/)
		{
			print WRITE "{\n";
			if ($#pattern_header_commets>0)
			{
				for (my $i=0; $i<=$#pattern_header_commets; $i++){print WRITE "$pattern_header_commets[$i]\n";}
			}
			next;
		}
		#4)process global subr label
		$originStr=~s/subr\s*// if($originStr=~/^global\s*subr\s*\w+:/i);#global subr -> global
		
#		#process label to combine with vector. due to label in a seperate line
#		if ($originStr=~/:$/){$label_line=$originStr.' '; next;}#label line combine to the next vector
#		$originStr=$label_line.$originStr;#vector combine with previous label
#		$label_line="";#reset label		
		
		#5)check if it's not valid vector
		goto WRITE_DIRECTLY if (!($originStr=~ />/));#valida vector has symbol ">"
		
		#5)check pattern Label and Opcode
		if ( $str_line =~ /^(?<label>.*:)?(?<opcode>.*)>.*;/ or $str_line =~ /^(?<label>\w+\s?:)?(?<opcode>.*)(?<comment>\/\/.*)?/)
		{
			$result = $+{opcode} if ( $+{opcode} );
			if($result ne "" and !($result=~/>/))#skip scan data
			{
				$tempt_opcode=$result;
				$tempt_opcode=~s/\s+$//;
				if($result=~/^repeat.*/){$tempt_opcode =~ s/\d+//;$tempt_opcode=~s/^.*://;}#only for repeat opcode process
			}

		}
		#5)process and convert J750 opcode to the UltraFlex opcode
		if ( $result =~ /mrepeat/ ) {$originStr =~ s/mrepeat/repeat/;} #mrepeat -> repeat
		elsif($result=~/end_module/) {$originStr =~ s/end_module/halt/;} #end_module -> halt
		elsif($result=~/ign/) {$originStr =~ s/\bign\b/mask/;} #ign -> mask
		elsif($result=~/clr_fail/) {$originStr =~ s/clr_fail//; $originStr=$originStr."\/\/clr_fail in J750";} #clr_fail move to comments
		elsif($result=~/set_code/) {$originStr=~/^(set_code\s\d.*)>/; my $match=$1; $match=~s/\s+$//; $originStr =~ s/^.*>/ >/; $originStr=$originStr."//".$match." in J750";} #set_code change to comments
		elsif($result=~/clr_code/) {$originStr =~ s/^.*>/ >/; $originStr=$originStr."\/\/clr_code in J750";} #set_code change to comments
		elsif($result=~/if\s*\(flag\)\s*jump/) {$originStr =~ s/\bflag\b/branch_expr/;} #flag -> branch_expr
		elsif($result=~/clr_flag\s*\(cpuA\)/) {$originStr =~ s/clr_flag/clr_cond_flags/; $originStr=~s/cpuA/cpuA_cond/;}
		elsif($result=~/clr_flag\s*\(fail\)/) {$originStr =~ s/clr_flag/clr_cond_flags/;}
#		elsif($result =~ /set_cpu\s*\(cpuA\)/ ) {$originStr =~ s/set_cpu/set_cpu_cond/; $originStr =~ s/cpuA/cpuA_cond/;}
		elsif($result =~ /exit_loop/ ) {$originStr =~ s/exit_loop/clr_loop/;}
		elsif($result =~ /push/ ) {$originStr =~ s/\bpush\b/push_subr/;}
		elsif($result =~ /call_glo/ ) {$originStr =~ s/call_glo/call globalAddr/;}
		elsif(($result=~/enable/)&&($result=~/none/)) {$originStr =~ s/enable/branch_expr=/;} #enable -> branch_expr		
		elsif($result=~/enable/) {
			$originStr =~ s/enable/branch_expr=/;
			$originStr=~s/cpuA/cpuA_cond/i;
			$originStr=~s/cpuB/cpuB_cond/i;
			$originStr=~s/cpuC/cpuC_cond/i;
			$originStr=~s/cpuD/cpuD_cond/i;			
		} #enable -> branch_expr
		#======process cpuFlag, remove the J750 cpu opcode to comments
		elsif($result=~/cpu/)
		{
			print WRITE "//	TER2018: below vector has opcode: $result\n";#add comments where will remove opcode
			$originStr=substr($originStr,length($result)); #remove the opcode command			
		}
		elsif($result =~ m/\bloop\w\s+(\d+)/ or $result =~ m/set_loop\w\s+(\d+)/) #loop or set_loop... #expand the loop command
		{
			my $loop_count=$1;#get loop counts
			my @loopcontent;
			print WRITE "// TER: Here loop start by loop_count $loop_count\n";
			$originStr=~s/^.*>/ >/;#remove opcode
			push(@loopcontent,$originStr);
			while($originStr=<READ>)
			{
				$linenumber++;
				$originStr=~s/^\s+|\s+$//;
				$originStr='//label '.$originStr.' in J750' if($originStr=~/^\w+:$/);#comment the Label
				last if ($originStr=~/^end_loop\w\s/i);#special case like control bit: end_loopA Start, stv > tset
				push(@loopcontent,$originStr);
			}
			my $stv_exist="";
			$stv_exist="stv" if ($originStr=~/\bstv/);
			$originStr=~s/^.*>/ >/;#remove opcode and control bit
			$originStr=$stv_exist.$originStr; #setback stv if vectort has stv control bit
			push(@loopcontent,$originStr);
			for (my $looping=0;$looping< $loop_count;$looping++)#expand the loop command			
			{
				for(my $i=0;$i<($#loopcontent+1);$i++) #print vectors
				{
					print WRITE "// TER: Here loop end by loop_count $loop_count\n" if (($looping== $loop_count-1) and ($i==$#loopcontent));
					print WRITE " $loopcontent[$i] \n";
				}
			}
			next;#jump to next read line
		} #elsif($result =~ m/\bloop\w\s+(\d+)/ or $result =~ m/set_loop\w\s+(\d+)/) #loop or set_loop... #expand the loop command
		else{} #nothing to do
		#===========save opcode
		if ($tempt_opcode ne "")
		{
			my $exist_cnt=(grep /^$tempt_opcode$/, @opcodelist);
			if ($exist_cnt==0)
			{
				push( @opcodelist, $tempt_opcode );
				print WRITE_opcode_location "$linenumber:  ";
				print WRITE_opcode_location "$orginal_j750\n";
				print WRITE_opcode "$tempt_opcode\n";				
			}			
		}					
		#nothing to process, just write original context directly
		WRITE_DIRECTLY:	print (WRITE "$originStr\n");
	}
	
	#close files
	close READ;
	close WRITE;
	close READTEMPT;
}

sub UltraFlexPatternCompiler()
{
	my $input_file=$_[0];#"UltraFlex_SRM_VM_Dummy.atp ";
	#$output="-output ";
	my $pinmap="-pinmap_workbook ";
	my $digital_inst="-digital_inst ";
	my $opcode_mode="-opcode_mode ";
	my $scan_type="-scan_type ";
	my $save_comments="-comments ";
	
	$pinmap=$pinmap . "Pinmap.txt "; #specified the pinmap file
	$digital_inst=$digital_inst . "HSDMQ "; #specified the digital instrument
	$opcode_mode=$opcode_mode ."single "; #specififed the timing mode
	$scan_type=$scan_type."x2 " if $is_scan_pat;#specified the scan type in 2x for scan pattern
	
	my $switches=$pinmap.$digital_inst.$opcode_mode;
	if($is_scan_pat){$switches=$switches.$scan_type.$save_comments;	}
	else			{$switches=$switches.$save_comments;}
	
	my $my_command="apc ".$input_file ." " .$switches;
	
	#system("apc -stdin -output bar.pat");
	system($my_command);#invoke the pattern compiler by command-line interface: apc input-ascii-file(s) [switches]
}

File: J750_to_Ultra_compiler.py
#!/usr/bin/python
# -*- coding: UTF-8 -*-

import os,re,sys

dir=r'D:\Project_Test\Python\J750_to_Ultra_compiler'
make_dir=r'D:\Project_Test\Python\J750_to_Ultra_compiler\UltraFlex'
J750_dir=r'D:\Project_Test\Python\J750_to_Ultra_compiler\J750_Pattern'
opcode_list=r'D:\Project_Test\Python\J750_to_Ultra_compiler\opcode_list.txt'
opcode_in_pattern=r'D:\Project_Test\Python\J750_to_Ultra_compiler\opcode_in_pattern.txt'
#os.rmdir(make_dir)

if os.path.exists(make_dir):
    print("The folder already exists!")
else:
    os.mkdir(make_dir,777)  #mode=777  if mode=0777 will be error
    print("The folder created successfully!")

if os.path.exists(J750_dir):
    os.chdir(J750_dir)
#use for save opcode and opcode located in pattern
opcode_list_fp=open(opcode_list,'w')
opcode_in_pattern_fp=open(opcode_in_pattern,'w')

#traversal the J750 Folder
J750_atp_list=os.listdir(J750_dir)
def findopcode():
    reapt_flag = 0
    halt_flag = 0
    scan_setup_flag = 0
    line_no = 0  # line number
    atp_file_name_flag = 0
    vector_num_flag=0
    bracket_flag=0
    reg=''
    for atp_file in J750_atp_list:
        vector_num_flag = 0
        UltraAtpPath = os.path.join(os.path.abspath('..'), 'UltraFlex', atp_file)
        if re.match(r'.*\.atp', atp_file):
            print(atp_file)
            vector_num = -1  # statistics the vector num
            for vector_num, line in enumerate(open(os.path.join(J750_dir, atp_file), 'rU')):
                pass
                vector_num += 1
            with open(os.path.join(J750_dir, atp_file), 'r') as fp:
                UltraAtpPath_fp = open(UltraAtpPath, 'w+')
                lines=[]    #insert at the specofied line
                for line in UltraAtpPath_fp:
                    lines.append(line)
                atp_line = fp.readline()  # readline读完本行就会自动定位到下一行  一直读到没有下一行
                while atp_line:
                    line_no = line_no + 1
                    if re.match(r'import tset\s+[\d\w\s]', atp_line):
                        #get the string after the import tset---start
                        tset_label=re.sub(r'import tset\s+','',atp_line)
                        tset_label = re.sub(r';', '', tset_label)
                        tset_label=tset_label.replace('\n','')#delet \n
                        reg = re.compile(r'\s+>\s+' + tset_label)
                        # get the string after the import tset---end
                        lines.insert(0, atp_line)  # insert at the first row
                        s = ''.join(lines)
                        UltraAtpPath_fp = open(UltraAtpPath, 'w')
                        UltraAtpPath_fp.write(s)
                    if vector_num_flag == 0:
                        atp_file = re.sub(r'\.atp$', '',atp_file)
                        if (vector_num > 64):
                            lines.insert(1, 'vm_vector ' + atp_file + '\n')  # insert at the second row
                        else:
                            lines.insert(1, 'vm_vector ' + atp_file + '\n')
                        s = ''.join(lines)
                        UltraAtpPath_fp = open(UltraAtpPath, 'w')
                        UltraAtpPath_fp.write(s)
                        vector_num_flag = 1
                    if re.match(r'\s*vector\s+\(\s*\$tset\s*\,', atp_line):
                        pin_line = re.sub(r'\s*vector\s+\(\s*\$tset\s*\,', '($tset ', atp_line)
                        lines.insert(2, pin_line)  # insert at the first row
                        s = ''.join(lines)
                        UltraAtpPath_fp = open(UltraAtpPath, 'w')
                        UltraAtpPath_fp.write(s)
                        # UltraAtpPath_fp.write(pin_line)
                        if bracket_flag == 0:
                            lines.insert(3, '{\n')  # insert at the first row
                            s = ''.join(lines)
                            UltraAtpPath_fp = open(UltraAtpPath, 'w')
                            UltraAtpPath_fp.write(s)
                            bracket_flag = 1
                    if re.match(r'//', atp_line):
                        lines.insert(5, atp_line)
                        s = ''.join(lines)
                        UltraAtpPath_fp = open(UltraAtpPath, 'w')
                        UltraAtpPath_fp.write(s)
                    if re.match(r'start_label\s+', atp_line):
                        UltraAtpPath_fp.write(atp_line)
                    '''
                    write opcode_list&opcode_in_pattern&UltraAtpPath-----Start!!!
                    '''
#match repeat
                    if re.match(r'\s*repeat\s+\d+\s+\>', atp_line):
                        if reapt_flag == 0:
                            reapt_flag = 1
                            opcode_list_fp.write('repeat\n')#write into opcode_list file
                            opcode_in_pattern_fp.write(
                                str(line_no) + ":" + " " + atp_line)  # String concatenation and output
                        UltraAtpPath_fp.write(atp_line)
#match halt
                    if re.match(r'\s*halt\s+\>\s*', atp_line):
                        if halt_flag == 0:
                            halt_flag = 1
                            opcode_list_fp.write('halt\n')  # write opcode to opcode_list
                            opcode_in_pattern_fp.write(
                                str(line_no) + ":" + " " + atp_line)  # write opcode and pattern to opcode_in_pattern
                        UltraAtpPath_fp.write(atp_line)
#match sacn_setup
                    if re.match(r'\s*scan_setup\s+\>\s*', atp_line):
                        if scan_setup_flag == 0:
                            scan_setup_flag = 1
                            opcode_list_fp.write('scan_setup\n')
                            opcode_in_pattern_fp.write(str(line_no) + ":" + " " + atp_line)
                        UltraAtpPath_fp.write(atp_line)
#no opcode,match tset name
                    if re.match(reg,atp_line):
                        modify_line=re.sub(r'\s+>\s+','> ',atp_line)
                        UltraAtpPath_fp.write(modify_line)
#match /* multi-line matching
                    '''
                    if re.match(r'^/\*\s*',atp_line):
                        UltraAtpPath_fp.write(atp_line)
                        atp_line = fp.readline()
                        while re.match(r'\*/',atp_line) == None:
                            UltraAtpPath_fp.write(atp_line)
                            atp_line = fp.readline()
                        else:
                            UltraAtpPath_fp.write(atp_line)
                    '''
                    if (0==atp_line.find('/*'))&(0==atp_line.find('*/')):
                        UltraAtpPath_fp.write(atp_line)
                    elif (0==atp_line.find('/*'))&(-1==atp_line.find('*/')):
                        UltraAtpPath_fp.write(atp_line)
                        atp_line = fp.readline()
                        while atp_line.find('*/') == -1:
                            UltraAtpPath_fp.write(atp_line)
                            atp_line = fp.readline()
                        else:
                            UltraAtpPath_fp.write(atp_line)
#match //
                    if re.match(r'//\s*', atp_line):
                        UltraAtpPath_fp.write(atp_line)

                    if re.match(r'^\n',atp_line):
                        re.sub(r'\n', '}', atp_line)
                        UltraAtpPath_fp.write(atp_line)
                    '''
                    write opcode_list&opcode_in_pattern&UltraAtpPath-----End!!!
                    '''
                    if atp_file_name_flag == 0:
                        opcode_in_pattern_fp.write("--------------------" + atp_file + "--------------------" + '\n')
                        atp_file_name_flag = 1
                    atp_line = fp.readline()
                UltraAtpPath_fp.write('}')
                atp_file_name_flag = 0


def ConvertPat():
    vector_num_flag=0
    bracket_flag=0
    for atp_file in J750_atp_list:
        UltraAtpPath=os.path.join(os.path.abspath('..'),'UltraFlex',atp_file)
        if re.match(r'.*\.atp', atp_file):
            vector_num = -1#statistics the vector num
            for vector_num, line in enumerate(open(os.path.join(J750_dir, atp_file), 'rU')):
                pass
                vector_num += 1
            with open(os.path.join(J750_dir, atp_file), 'r') as fp:
                UltraAtpPath_fp = open(UltraAtpPath, 'w+')
                lines=[]    #insert at the specofied line
                for line in UltraAtpPath_fp:
                    lines.append(line)
                atp_line = fp.readline()  # readline读完本行就会自动定位到下一行  一直读到没有下一行
                while atp_line:
                    if re.match(r'import tset\s+[\d\w\s]',atp_line):
                        lines.insert(0,atp_line)  #insert at the first row
                        s=''.join(lines)
                        UltraAtpPath_fp = open(UltraAtpPath, 'w')
                        UltraAtpPath_fp.write(s)
                    if vector_num_flag==0:
                        if (vector_num > 64):
                            lines.insert(1, 'vm_vector ' + atp_file+ '\n')  #insert at the second row
                        else:
                            lines.insert(1, 'vm_vector ' + atp_file+ '\n')
                        s = ''.join(lines)
                        UltraAtpPath_fp = open(UltraAtpPath, 'w')
                        UltraAtpPath_fp.write(s)
                        vector_num_flag=1
                    if re.match(r'\s*vector\s+\(\s*\$tset\s*\,',atp_line):
                        pin_line=re.sub(r'\s*vector\s+\(\s*\$tset\s*\,','($tset ',atp_line)
                        lines.insert(2,pin_line)  #insert at the first row
                        s=''.join(lines)
                        UltraAtpPath_fp = open(UltraAtpPath, 'w')
                        UltraAtpPath_fp.write(s)
                        #UltraAtpPath_fp.write(pin_line)
                        if bracket_flag==0:
                            lines.insert(3, '{\n')  # insert at the first row
                            s = ''.join(lines)
                            UltraAtpPath_fp = open(UltraAtpPath, 'w')
                            UltraAtpPath_fp.write(s)
                            bracket_flag=1
                    if re.match(r'//',atp_line):
                        lines.insert(5,atp_line)
                        s = ''.join(lines)
                        UltraAtpPath_fp = open(UltraAtpPath, 'w')
                        UltraAtpPath_fp.write(s)
                    if re.match(r'start_label\s+',atp_line):
                        UltraAtpPath_fp.write(atp_line)
                    atp_line = fp.readline()

def AtpToPat():
    print('Pat')
    print('')
    print('')
findopcode()
#ConvertPat()
print("Finished!")


